src/CodeGen_C.cpp

```diff
@@ -219,6 +219,60 @@ class HalideFreeHelper {
 } // namespace
 )INLINE_CODE";
 
+const string post_halide_runtime_globals = R"INLINE_CODE(
+namespace {
+void _halide_buffer_to_string(char* dst, size_t n, const halide_buffer_t *buf) {
```

Comment:
Is it not possible to just call the halide_buffer_to_string already in the runtime?

Reply:
There isn't any guarantee that that version will be available for use (e.g., if we are using the C backend because we can't target it via LLVM)

Reply:
Wait, does the C backend not require a runtime? I thought there were plenty of calls it could make into the runtime still.

Reply:
(After offline discussion)

You need a runtime to use the C backend (e.g. for halide_malloc!). People can either generate one in the standard way, though that raises the question of why they're using the C backend, or they can attempt to compile our runtime sources directly and link that in, or they can just implement runtime functions piecemeal as they need them. I assume the latter is happening here. I don't think we should start inlining duplicate implementations of our runtime functionality into the generated C, because it doesn't solve the problem of needing a runtime unless you implement everything (e.g. halide_do_par_for), and whatever we do implement is a piece of repeated functionality to maintain and test. So I think we should just emit a call to halide_buffer_to_string, and the downstream environment with a custom runtime should implement halide_buffer_to_string.

Reply:
Xtensa backend (from xtensa-codegen branch) is based on C backend, and we indeed compile and link runtime sources directly, so I agree that in this case we should also just emit a call and then link corresponding runtime piece.

Reply:
ok, I'll split this PR into several pieces